core/services/gas_updater.go

```diff
@@ -0,0 +1,140 @@
+package services
+
+import (
+	"fmt"
+	"math/big"
+	"sort"
+
+	"chainlink/core/eth"
+	"chainlink/core/logger"
+	"chainlink/core/store"
+	"chainlink/core/store/models"
+	"chainlink/core/utils"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var (
+	promGasUpdaterAllPercentiles = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "gas_updater_all_gas_percetiles",
+		Help: "Gas price at given percentile",
+	},
+		[]string{"percentile"},
+	)
+
+	promGasUpdaterSetGasPrice = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "gas_updater_set_gas_price",
+		Help: "Gas updater set gas price (in Wei)",
+	},
+		[]string{"percentile", "block_num"},
+	)
+)
+
+//go:generate mockery -name GasUpdater  -output ../internal/mocks/ -case=underscore
+
+// GasUpdater listens for new heads and updates the base gas price dynamically
+// based on the configured percentile of gas prices in that block
+type GasUpdater interface {
+	store.HeadTrackable
+	RollingBlockHistory() []eth.Block
+}
+
+type gasUpdater struct {
+	store                   *store.Store
+	rollingBlockHistory     []eth.Block
+	rollingBlockHistorySize int
+	blockDelay              int64
+	percentile              int
+}
+
+// NewGasUpdater returns a new gas updater.
+func NewGasUpdater(store *store.Store) GasUpdater {
+	gu := &gasUpdater{
+		store:                   store,
+		rollingBlockHistory:     make([]eth.Block, 0),
+		rollingBlockHistorySize: int(store.Config.GasUpdaterBlockHistorySize()),
+		blockDelay:              int64(store.Config.GasUpdaterBlockDelay()),
+		percentile:              int(store.Config.GasUpdaterTransactionPercentile()),
+	}
+	return gu
+}
+
+func (gu *gasUpdater) Connect(bn *models.Head) error {
+	if gu.store.Config.GasUpdaterEnabled() {
+		logger.Debugw("GasUpdater: dynamic gas updates are enabled", "ethGasPriceDefault", gu.store.Config.EthGasPriceDefault())
+	} else {
+		logger.Debugw("GasUpdater: dynamic gas updating is disabled", "ethGasPriceDefault", gu.store.Config.EthGasPriceDefault())
+	}
+	return nil
+}
+
+func (gu *gasUpdater) Disconnect() {
+	return
+}
+
+// OnNewHead recalculates and sets global gas price on every head
+func (gu *gasUpdater) OnNewHead(head *models.Head) {
+	if !gu.store.Config.GasUpdaterEnabled() {
+		return
+	}
+	blockToFetch := head.Number - gu.blockDelay
+	if blockToFetch < 0 {
+		logger.Warnf("GasUpdater: skipping gas calculation, current block height %v is lower than GAS_UPDATER_BLOCK_DELAY of %v", head.Number, gu.blockDelay)
+		return
+	}
+	blockNumber := utils.Uint64ToHex(uint64(blockToFetch))
+	block, err := gu.store.TxManager.GetBlockByNumber(blockNumber)
+	if err != nil {
+		logger.Error(err, fmt.Sprintf("GasUpdater: error retrieving block %v", blockToFetch))
+		return
+	}
+	logger.Debugw("GasUpdater: got block", "blockNumber", blockToFetch)
```

Comment:
Occasionally geth seems to return an "empty" block even when etherscan indicates that it has transactions.

I don't know why this happens.

Reply:
My guess is that a geth node gets the headers first via [`GetBlockHeaders`](https://github.com/ethereum/devp2p/blob/master/caps/eth.md#chain-synchronization), and makes a black entry for the missing blocks in its DB, which results in an empty `Transactions` list until it gets a response from `GetBlockBodies`.

It should be quite rare that miner economics results in an empty block, right? Wouldn't that only happen in the interval between the miner receiving a new block, and constructing a prospective block on top of that, which should happen in less than 1% of the time for a new block to appear.

Reply:
I checked. Occasionally miner economics really do result in empty blocks, but we end up seeing more empty blocks than this would warrant.

I suspect some tuning of the block delay is probably in order.